File: clients/java/src/main/java/spb_ai_champ/codegame/ClientMessage.java
package spb_ai_champ.codegame;

import spb_ai_champ.util.StreamUtil;

/**
 * Message sent from client
 */
public abstract class ClientMessage {
    /**
     * Write ClientMessage to output stream
     */
    public abstract void writeTo(java.io.OutputStream stream) throws java.io.IOException;

    /**
     * Read ClientMessage from input stream
     */
    public static ClientMessage readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
            case DebugMessage.TAG:
                return DebugMessage.readFrom(stream);
            case ActionMessage.TAG:
                return ActionMessage.readFrom(stream);
            case DebugUpdateDone.TAG:
                return DebugUpdateDone.readFrom(stream);
            case RequestDebugState.TAG:
                return RequestDebugState.readFrom(stream);
            default:
                throw new java.io.IOException("Unexpected tag value");
        }
    }

    /**
     * Ask app to perform new debug command
     */
    public static class DebugMessage extends ClientMessage {
        public static final int TAG = 0;
    
        /**
         * Command to perform
         */
        private spb_ai_champ.debugging.DebugCommand command;
    
        /**
         * Command to perform
         */
        public spb_ai_champ.debugging.DebugCommand getCommand() {
            return command;
        }
    
        /**
         * Command to perform
         */
        public void setCommand(spb_ai_champ.debugging.DebugCommand value) {
            this.command = value;
        }
    
        public DebugMessage(spb_ai_champ.debugging.DebugCommand command) {
            this.command = command;
        }
    
        /**
         * Read DebugMessage from input stream
         */
        public static DebugMessage readFrom(java.io.InputStream stream) throws java.io.IOException {
            spb_ai_champ.debugging.DebugCommand command;
            command = spb_ai_champ.debugging.DebugCommand.readFrom(stream);
            return new DebugMessage(command);
        }
    
        /**
         * Write DebugMessage to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            command.writeTo(stream);
        }
    
        /**
         * Get string representation of DebugMessage
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("DebugMessage { ");
            stringBuilder.append("command: ");
            stringBuilder.append(String.valueOf(command));
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }

    /**
     * Reply for ServerMessage::GetAction
     */
    public static class ActionMessage extends ClientMessage {
        public static final int TAG = 1;
    
        /**
         * Player's action
         */
        private spb_ai_champ.model.Action action;
    
        /**
         * Player's action
         */
        public spb_ai_champ.model.Action getAction() {
            return action;
        }
    
        /**
         * Player's action
         */
        public void setAction(spb_ai_champ.model.Action value) {
            this.action = value;
        }
    
        public ActionMessage(spb_ai_champ.model.Action action) {
            this.action = action;
        }
    
        /**
         * Read ActionMessage from input stream
         */
        public static ActionMessage readFrom(java.io.InputStream stream) throws java.io.IOException {
            spb_ai_champ.model.Action action;
            action = spb_ai_champ.model.Action.readFrom(stream);
            return new ActionMessage(action);
        }
    
        /**
         * Write ActionMessage to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            action.writeTo(stream);
        }
    
        /**
         * Get string representation of ActionMessage
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("ActionMessage { ");
            stringBuilder.append("action: ");
            stringBuilder.append(String.valueOf(action));
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }

    /**
     * Signifies finish of the debug update
     */
    public static class DebugUpdateDone extends ClientMessage {
        public static final int TAG = 2;
    
    
        public DebugUpdateDone() {
        }
    
        /**
         * Read DebugUpdateDone from input stream
         */
        public static DebugUpdateDone readFrom(java.io.InputStream stream) throws java.io.IOException {
            return new DebugUpdateDone();
        }
    
        /**
         * Write DebugUpdateDone to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
        }
    
        /**
         * Get string representation of DebugUpdateDone
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("DebugUpdateDone { ");
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }

    /**
     * Request debug state from the app
     */
    public static class RequestDebugState extends ClientMessage {
        public static final int TAG = 3;
    
    
        public RequestDebugState() {
        }
    
        /**
         * Read RequestDebugState from input stream
         */
        public static RequestDebugState readFrom(java.io.InputStream stream) throws java.io.IOException {
            return new RequestDebugState();
        }
    
        /**
         * Write RequestDebugState to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
        }
    
        /**
         * Get string representation of RequestDebugState
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("RequestDebugState { ");
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }
}

File: clients/kotlin/src/main/kotlin/spb_ai_champ/codegame/ClientMessage.kt
package spb_ai_champ.codegame

import spb_ai_champ.util.StreamUtil

/**
 * Message sent from client
 */
abstract class ClientMessage {
    /**
     * Write ClientMessage to output stream
     */
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)

    companion object {
        /**
         * Read ClientMessage from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): ClientMessage {
            when (StreamUtil.readInt(stream)) {
                DebugMessage.TAG -> return DebugMessage.readFrom(stream)
                ActionMessage.TAG -> return ActionMessage.readFrom(stream)
                DebugUpdateDone.TAG -> return DebugUpdateDone.readFrom(stream)
                RequestDebugState.TAG -> return RequestDebugState.readFrom(stream)
                else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }

    /**
     * Ask app to perform new debug command
     */
    class DebugMessage : ClientMessage {
        /**
         * Command to perform
         */
        var command: spb_ai_champ.debugging.DebugCommand
    
        constructor(command: spb_ai_champ.debugging.DebugCommand) {
            this.command = command
        }
    
        /**
         * Write DebugMessage to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            command.writeTo(stream)
        }
    
        /**
         * Get string representation of DebugMessage
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("DebugMessage { ")
            stringBuilder.append("command: ")
            stringBuilder.append(command)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 0
    
            /**
             * Read DebugMessage from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): DebugMessage {
                var command: spb_ai_champ.debugging.DebugCommand
                command = spb_ai_champ.debugging.DebugCommand.readFrom(stream)
                return DebugMessage(command)
            }
        }
    }

    /**
     * Reply for ServerMessage::GetAction
     */
    class ActionMessage : ClientMessage {
        /**
         * Player's action
         */
        var action: spb_ai_champ.model.Action
    
        constructor(action: spb_ai_champ.model.Action) {
            this.action = action
        }
    
        /**
         * Write ActionMessage to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            action.writeTo(stream)
        }
    
        /**
         * Get string representation of ActionMessage
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("ActionMessage { ")
            stringBuilder.append("action: ")
            stringBuilder.append(action)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 1
    
            /**
             * Read ActionMessage from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): ActionMessage {
                var action: spb_ai_champ.model.Action
                action = spb_ai_champ.model.Action.readFrom(stream)
                return ActionMessage(action)
            }
        }
    }

    /**
     * Signifies finish of the debug update
     */
    class DebugUpdateDone : ClientMessage {
    
        constructor() {
        }
    
        /**
         * Write DebugUpdateDone to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
        }
    
        /**
         * Get string representation of DebugUpdateDone
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("DebugUpdateDone { ")
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 2
    
            /**
             * Read DebugUpdateDone from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): DebugUpdateDone {
                return DebugUpdateDone()
            }
        }
    }

    /**
     * Request debug state from the app
     */
    class RequestDebugState : ClientMessage {
    
        constructor() {
        }
    
        /**
         * Write RequestDebugState to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
        }
    
        /**
         * Get string representation of RequestDebugState
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("RequestDebugState { ")
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 3
    
            /**
             * Read RequestDebugState from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): RequestDebugState {
                return RequestDebugState()
            }
        }
    }
}